### ISSUE TITLE
making stack work with ghcmod-vim?

### ISSUE THREAD

**Issue by timchu** [eagletmt/ghcmod-vim#88]
Somewhat new to ghcmod-vim.
I installed ghcmod with stack, and tried to do a :GhcModType on a function in a stack project, and it keeps telling me that 

ghcmod#command#type: Cannot guess type

Even though my project builds fine in stack. I'm somewhat new here, so I'm not sure whether it's a problem with my ghcmod, with my PATH, or with my ghcmod-vim.
All the best

**Comment by igrep** [comment]
Have you executed `ghc-mod YourFile.hs` directly at your project directory?
ghc-mod started to support stack from ver. 5.4.0.0.

https://mail.haskell.org/pipermail/haskell-cafe/2015-September/121412.html

**Comment by DanielG** [comment]
It's `ghc-mod check YourFile.hs` actually.

**Comment by igrep** [comment]
oh, sorry.

**Comment by davidpdrsn** [comment]
I'm having the same issue. Running `ghc-mod type src/Main.hs` from the command line works just fine, but running `:GhcModType` from within Vim gives me that error.

I'm using a ghc-mod that I compiled myself today.

**Comment by lobachevzky** [comment]
Same here. One thing I've noticed is that ghc-mod commands works from the command line when I run `stack exec ghc-mod`... I'm guessing that ghcmod-vim is running ghc-mod globally, not from within the stack environment. Is there a way to fix that?

**Comment by reiddraper** [comment]
Using ghc-mod from post Dec. 5, you'll need something like this https://github.com/reiddraper/ghcmod-vim/commit/f39dd594081d4095ad4e217e6e5c01ece59efbb3.

ghc-mod no longer accepts the `MODULE` argument to its commands, which ghcmod-vim was still providing.

**Comment by Tehnix** [comment]
Were these changes ever merged? I at least still seem to get`ghcmod#command#type: Cannot guess type` with stack.

**Comment by silky** [comment]
i'm getting this error on a stack project with docker

it seems that ghc-mod is just using the `stack build ...` when it needs to use my special docker env. `stack --docker --docker-image=... build`

it'd be nice if it could read the appropriate command somehow.

**Comment by unode** [comment]
I'm also still seeing this error. Has this project stagnated?
Last commit is almost a year and half old.
Running `ghc-mod type` manually works fine.

**Comment by unode** [comment]
Turns out my case was due to https://github.com/DanielG/ghc-mod/issues/883 .

**Comment by unode** [comment]
https://github.com/DanielG/ghc-mod/issues/883 magically fixed itself which allowed me to identify one more issue.

In one of the cases where this error happens the `.hs` file being checked lacks a `main` function and is not declared as a module.
In this case `ghc-mod` would print `EXCEPTION: types: The IO action ‘main’ is not defined in module ‘Main’` but this message is hidden due to the use of `--silent`. `ghcmod-vim` doesn't seem to handle these either as removing `--silent` from the command arguments produced no visible difference inside vim.

**Comment by BjornMelgaard** [comment]
Hi, I have ghc-mod installed by stack and `.local/bin` in `PATH`
What I observe is that ghc-mod cant find global stack libraries, unless called via `stack exec`

```
$ ghc-mod debug
ghc-mod: ghc: readCreateProcess: runInteractiveProcess: exec: does not exist (No such file or directory)
$ stack exec ghc-mod debug
Version:              ghc-mod-5.7.0.0
Library GHC Version:  8.0.2
System GHC Version:   8.0.2
Root directory:       /home/bjorn/projects/haskell-programming-from-first-principles/chapter26
Current directory:    /home/bjorn/projects/haskell-programming-from-first-principles/chapter26
GHC Package flags:
    -i/home/bjorn/projects/haskell-programming-from-first-principles/chapter26
    -i/home/bjorn/projects/haskell-programming-from-first-principles/chapter26
    -global-package-db -user-package-db -Wall
GHC System libraries: /home/bjorn/.stack/programs/x86_64-linux/ghc-8.0.2/lib/ghc-8.0.2
$ whereis ghc-mod
ghc-mod: /home/bjorn/.local/bin/ghc-mod 
```

**Comment by BjornMelgaard** [comment]
Hi again,  I solved this issue with wrapper script and `g:ghcmod_use_basedir` :smile: 

```sh
$ cat /home/bjorn/.config/dotfiles/bin/ghc-mod
#!/bin/sh
stack exec ghc-mod "$@"
```
```vim
"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
" => Haskell
"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
call dein#add('neovimhaskell/haskell-vim')
call dein#add('eagletmt/neco-ghc')
autocmd FileType haskell setlocal omnifunc=necoghc#omnifunc
call dein#add('Shougo/vimproc.vim', {'build' : 'make'})
call dein#add('eagletmt/ghcmod-vim')
let g:ghcmod_use_basedir="/home/bjorn/.config/dotfiles/bin"
```

**Comment by BjornMelgaard** [comment]
this variant is more accurate
```
#!/bin/sh
stack exec ghc-mod -- "$@"
```

**Comment by madnight** [comment]
By the way does anyone know if this repo is dead? Last commit is over a year ago, many open issues and unreviewed pull requests.

**Comment by ecly** [comment]
@BjornMelgaard I've tried to get it working using your script, but run into the issue: 
`ghcmod: ghc-mod is not exectuable!` when I open an .hs file in vim.

The script is a 1:1 clone of yours:

```
$ cat /home/ecly/Scripts/ghc-mod  
#!/bin/sh  
stack exec ghc-mod -- "$@"
```

The script file itself has every permission one could dream of. I have ghcmod-vim installed through vundle.

```
...
Plugin 'shiena/ghcmod-vim'
call vundle#end()            " required
let g:ghcmod_use_basedir="/home/ecly/Scripts/"
```

Running stack exec ghc-mod manually works as expected:
```
stack exec ghc-mod -- --version
ghc-mod version 5.7.0.0 compiled by GHC 8.0.2
```

Any idea?

**Comment by BjornMelgaard** [comment]
@ecly, yeah, `Directory with script must be before stack's $HOME/.local/bin`

like this 

```
 ~/projects/nextjs-typescript-boilerplate   master ●  echo $PATH | tr ':' '\n'
/home/bjorn/.rvm/gems/ruby-2.4.1/bin
/home/bjorn/.rvm/gems/ruby-2.4.1@global/bin
/home/bjorn/.rvm/rubies/ruby-2.4.1/bin
/home/bjorn/.bin # there is my wrapper, its link to  /home/bjorn/.config/dotfiles/bin/ghc-mod
/home/bjorn/projects/umsf/bin
/home/bjorn/.local/bin # there is ghc-mod from stack
/usr/local/heroku/bin
/usr/local/sbin
/usr/local/bin
/usr/bin
/usr/lib/jvm/default/bin
/usr/bin/site_perl
/usr/bin/vendor_perl
/usr/bin/core_perl
/home/bjorn/.node_modules/bin
/home/bjorn/.rvm/bin
 ~/projects/nextjs-typescript-boilerplate   refactoring  whereis ghc-mod
ghc-mod: /home/bjorn/.config/dotfiles/bin/ghc-mod /home/bjorn/.local/bin/ghc-mod

```

**Comment by ecly** [comment]
@BjornMelgaard 
Ahhh, I didn't realize I had to add it to my path as well! I thought basedir change was enough.
I added it in .vimrc and made it work:
`let $PATH = $PATH . ':' . expand('~/Scripts/bin')`
Thanks!

**Comment by BjornMelgaard** [comment]
awesome)) :sunglasses:
